chore(compass-aggregations): attach disabled stages as leading comments to enabled stages COMPASS-6246

gribnoysup opened the pull request:
The reason we were not attaching comments correctly is that the logic would short circuit immediately when stage is enabled, leading to some situations ending up with most of the stages ending up attached to the last stage.

This patch fixes the issue by refactoring the logic a bit, instead of attaching comments as trailing, we do it similar to the babel parser where everything is first attached as a leading comment, and only the leftover comments are passed as trailing comments to the last stage

mabaasit commented:
Great! This solves the reported problem. 
However while testing it (the same pipeline), i ran into another issue of parsing. Comment out `$project` in text mode and toggle to builder mode. Notice `$project` stage does not show up at all. A follow up for this?

gribnoysup commented:
Just noticed that we don't really validate the statges in the tests, will include a fix and more test cases in this PR

gribnoysup commented:
Updated the code, found a few other corner-cases that I covered while at it and added them to the tests. @mabaasit if you have other cases that you usually test with but they are not in the test suite, can you share them so I can add them to the tests too?

gribnoysup commented:
Whoops, got to excited with test passing and merged. This is behind a feature flag though, so should be okay, I can address any other comments in the follow-up
